Add code allowing calculation of EnsoGL stats summaries

akavel opened the pull request:
### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

This change adds utility code for calculating summaries from multiple samples (snapshots) of EnsoGL runtime stats values.

This internal feature is expected to be used by Enso IDE performance profiling tools, which are planned to be added in the near future.

https://www.pivotaltracker.com/story/show/181093920

#### Visuals

A demo scene named `stats` was added, showcasing how to perform calculations using the new tools. Currently, the summary calculations in the scene work only when the EnsoGL stats Monitor Panel is visible; this is planned to be improved in a future task (https://www.pivotaltracker.com/story/show/181093601).



https://user-images.githubusercontent.com/273837/152835466-b56dbd8b-b802-405f-8b43-9a97444238c5.mov





### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

 - Note: the stats aggregation code is intended to be later used in Enso IDE's main rendering loop, so it needs to have very good performance characteristics. 
     - Due to that, `Accumulator` was designed to only use simple addition arithmetic, and be constant-memory once created.

[ci no changelog needed]

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/develop/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/develop/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

akavel committed:
initial raw port of relevant parts of old PR

akavel committed:
bring back frame_counter(), cannot remove it yet

akavel committed:
tweak comments and names

akavel committed:
reduce usage of allow(missing_docs)

akavel committed:
add overlooked Copy traits

akavel committed:
first sketch of a test; adding pub for fields

akavel committed:
more test

akavel committed:
starting to use macro in tests to reduce repetitiveness

akavel committed:
convert further tests to macro

akavel committed:
ValueAccumulator fields seem to not need to be pub

akavel committed:
drop serde for now

akavel committed:
improve readability in test

akavel committed:
small tweak to comments

akavel committed:
Merge branch 'develop' into wip/akavel/stats-summaries-181093920

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -177,3 +248,109 @@ macro_rules! if_compiled_with_stats {
+
+
+    macro_rules! test_with_new_sample {
+        ($accumulator:expr; $($field:ident : $type:tt = $sample:literal
+        => min: $min:literal avg: $avg:literal max: $max:literal)*) => {
```
Please format this code. In such complex cases, just make it multiline.

wdanilo commented on lib/rust/ensogl/core/src/debug/stats.rs:
```diff
@@ -177,3 +248,109 @@ macro_rules! if_compiled_with_stats {
+    }
+
+    #[test]
+    fn stats_summaries() {
+        // This tests attempts to verify calculation of proper summaries for stats of each
+        // primitive type supported by `gen_stats!`.
+
+        let mut accumulator: Accumulator = default();
+        assert!(matches!(accumulator.summarize(), None));
+
+        test_with_new_sample!(accumulator;
+            fps:               f64   = 55.0 => min: 55.0 avg: 55.0   max: 55.0
+            wasm_memory_usage: u32   = 1000 => min: 1000 avg: 1000.0 max: 1000
+            buffer_count:      usize = 3    => min: 3    avg: 3.0    max: 3
+        );
+        test_with_new_sample!(accumulator;
+            fps:               f64   = 57.0 => min: 55.0 avg: 56.0   max: 57.0
+            wasm_memory_usage: u32   = 2000 => min: 1000 avg: 1500.0 max: 2000
+            buffer_count:      usize = 2    => min: 2    avg: 2.5    max: 3
+        );
+        test_with_new_sample!(accumulator;
+            fps:               f64   = 56.0 => min: 55.0 avg: 56.0   max: 57.0
+            wasm_memory_usage: u32   = 3000 => min: 1000 avg: 2000.0 max: 3000
+            buffer_count:      usize = 1    => min: 1    avg: 2.0    max: 3
+        );
```
that's a really nice way of doing tests!

wdanilo reviewed (changes requested):
The code is 100% correct, however, nothing is using this code. I have always a big trouble when merging dead code, so I'd prefer not to merge it until this PR introduces anything that actually uses this code.

akavel committed:
review: WIP start adding new demo scene "stats"

akavel committed:
WIP show simple label

akavel committed:
WIP modify label text in loop

akavel committed:
WIP show FPS

akavel committed:
WIP show fps summary in demo scene

akavel committed:
WIP update FPS only once per 60 frames to speed things up

akavel committed:
tweak TODOs in demo scene

akavel committed:
rename things in demo scene

akavel committed:
review: make macro pattern multiline

akavel committed:
improve var names and delete comments in macro

akavel committed:
Merge remote-tracking branch 'origin/develop' into wip/akavel/stats-summaries-181093920

akavel committed:
calculate always, but print only every 60 frames

akavel committed:
add missing "the"

akavel commented:
@wdanilo I added a demo scene: it showcases the feature and provides an example of how the summarization API is intended to be used. I added a video recording of the demo scene to the PR. As mentioned by @MichaelMauderer in chat, those utilities are planned to be further used soon in https://www.pivotaltracker.com/story/show/181151201, which is waiting for them to be completed. Would the above unblock this PR for potential merging?

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/181151201">Integrate EnsoGL stats with profiling framework - Pivotal Tracker</a></strong></div></blockquote>

wdanilo reviewed (approved):
The new demo scene is amazing!

akavel commented:
(*Process note:* the code-review is now accepted, therefore the PR is waiting for Review by @MichaelMauderer on Task https://www.pivotaltracker.com/story/show/181093920 before merge.)

<blockquote><div><strong><a href="https://www.pivotaltracker.com/story/show/181093920">Add code allowing calculation of EnsoGL stats summaries - Pivotal Tracker</a></strong></div></blockquote>

akavel committed:
Merge branch 'develop' into wip/akavel/stats-summaries-181093920

akavel committed:
fix demo scene's doc comment
